libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableArray.m

```diff
@@ -0,0 +1,223 @@
+ and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright notice, this list of
+ conditions and the following disclaimer in the documentation and/or other materials provided
+ with the distribution.
+ * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
+ endorse or promote products derived from this software without specific prior written
+ permission of salesforce.com, inc.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "SFSDKSafeMutableArray.h"
+
+@interface SFSDKSafeMutableArray()
+@property (nonatomic,strong) NSMutableArray *backingArray;
+@property (nonatomic,strong) dispatch_queue_t queue;
+@end
+
+@implementation SFSDKSafeMutableArray
+
+- (id)init {
+    if ((self = [super init]))
+    {
+        self.backingArray = [NSMutableArray array];
+        [self initQueue];
+    }
+    return self;
+}
+
+- (id)initWithCapacity:(NSUInteger)numItems {
+    if ((self = [super init]))
+    {
+        self.backingArray = [NSMutableArray arrayWithCapacity:numItems];
+        [self initQueue];
+    }
+    return self;
+}
+
+- (NSUInteger)count {
+    __block NSUInteger size;
+    dispatch_sync(self.queue, ^{
+        size = self.backingArray.count;
+    });
+
+    return size;
+}
+
+-(BOOL)containsObject:(id)anObject {
+    __block BOOL exists;
+    dispatch_sync(self.queue, ^{
+        exists = [self.backingArray containsObject:anObject];
+    });
+    return exists;
+}
+
+-(id)objectAtIndexedSubscript:(NSUInteger)idx {
+    __block id object;
+    dispatch_sync(self.queue, ^{
+        object = [self.backingArray objectAtIndexedSubscript:idx];
+    });
+    return object;
+}
+
+-(id)objectAtIndexed:(NSUInteger)idx {
+    __block id object;
+    dispatch_sync(self.queue, ^{
+        object = [self.backingArray objectAtIndex:idx];
+    });
+    return object;
+}
+
+- (NSArray *)asArray {
+    __block NSArray *array;
+    dispatch_sync(self.queue, ^{
+        array = [NSArray arrayWithArray:self.backingArray];
+    });
+    return array;
+}
+
+- (void)filterUsingPredicate:(NSPredicate *)predicate {
+    dispatch_sync(self.queue, ^{
+        [self.backingArray filterUsingPredicate:predicate];
+    });
```

Comment:
Doesnt this mutate the backing array? I think it should use the barrier.

Reply:
oh let me update it.